### ISSUE TITLE
read_10x_h5 error in scanpy 1.9.1

### ISSUE THREAD

**Issue by francois-a** [broadinstitute/CellBender#128]
Hi, I'm not sure if this is related to other h5 issues (and may be a scanpy bug), but scanpy 1.9.1 throws `ValueError: Illegal slicing argument for scalar dataspace` when opening `*.cellbender_output_filtered.h5` files (this works fine with scanpy 1.8.2).

**Comment by sjfleming** [comment]
François!  Hope all is well!

So, I think I know what this is... any chance the input data was in CellRanger v2 format?  CellBender outputs a file in "v2 format" if the input was v2 format.  I guess I wanted to keep output like input, although this may have been a bad choice in hindsight, since v3 format from CellRanger makes a lot more sense in terms of its structure, and is better-maintained by writers of other tools.

If that's what's going on, it is a scanpy bug, which I am attempting to fix here
https://github.com/scverse/scanpy/pull/2248

**Comment by sjfleming** [comment]
If you are seeing that with a CellRanger v3 formatted output file from CellBender, let me know!

**Comment by dnjst** [comment]
I'm getting this error with CellRanger v3 formatted output

`~/project/conda/notebook_env3_clone1/lib/python3.8/site-packages/scanpy/readwrite.py in read_10x_h5(filename, genome, gex_only, backup_url)
    181         v3 = '/matrix' in f
    182     if v3:
--> 183         adata = _read_v3_10x_h5(filename, start=start)
    184         if genome:
    185             if genome not in adata.var['genome'].values:

~/project/conda/notebook_env3_clone1/lib/python3.8/site-packages/scanpy/readwrite.py in _read_v3_10x_h5(filename, start)
    266         try:
    267             dsets = {}
--> 268             _collect_datasets(dsets, f["matrix"])
    269 
    270             from scipy.sparse import csr_matrix

~/project/conda/notebook_env3_clone1/lib/python3.8/site-packages/scanpy/readwrite.py in _collect_datasets(dsets, group)
    254     for k, v in group.items():
    255         if isinstance(v, h5py.Dataset):
--> 256             dsets[k] = v[:]
    257         else:
    258             _collect_datasets(dsets, v)

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

~/project/conda/notebook_env3_clone1/lib/python3.8/site-packages/h5py/_hl/dataset.py in __getitem__(self, args, new_dtype)
    798         if self.shape == ():
    799             fspace = self.id.get_space()
--> 800             selection = sel2.select_read(fspace, args)
    801             if selection.mshape is None:
    802                 arr = numpy.zeros((), dtype=new_dtype)

~/project/conda/notebook_env3_clone1/lib/python3.8/site-packages/h5py/_hl/selections2.py in select_read(fspace, args)
     99     """
    100     if fspace.shape == ():
--> 101         return ScalarReadSelection(fspace, args)
    102 
    103     raise NotImplementedError()

~/project/conda/notebook_env3_clone1/lib/python3.8/site-packages/h5py/_hl/selections2.py in __init__(self, fspace, args)
     84             self.mshape = ()
     85         else:
---> 86             raise ValueError("Illegal slicing argument for scalar dataspace")
     87 
     88         self.mspace = h5s.create(h5s.SCALAR)

ValueError: Illegal slicing argument for scalar dataspace`

**Comment by sjfleming** [comment]
I think it's scalars and strings in h5 fields, since scanpy uses h5py to read h5 files using `[:]` notation (see here https://github.com/h5py/h5py/issues/1779)

I guess I could make sure everything that gets written is wrapped up into a list... sort of strange to have to do that, but it seems faster than hoping `scanpy` will change their dataloader.

For now, for existing files, here is a workaround:

1. Install `pytables` using `pip install pytables`
2. Use the utility `ptrepack` as follows to create a new h5 file that can be loaded by scanpy

```console
ptrepack CELLBENDER_OUTPUT.h5:/matrix MODIFIED_CELLBENDER_OUTPUT.h5:/matrix
```

This takes just the "matrix" part of the cellbender output file and puts it in its own new h5 file, without any of the cellbender metadata.  That new file should be readable directly by scanpy.

**Comment by esrice** [comment]
FYI, I tried the `ptrepack` solution but am sadly still getting the same error (`Illegal slicing argument for scalar dataspace`) when trying to read the ptrepack output file with scanpy's `read_10x_h5()`.

**Comment by sjfleming** [comment]
@esrice Thanks for letting me know.  That is confusing to me... there must be something going on that I'm not understanding.  Is it possible for you to share that `.h5` file here so I can take a look and see what's going on with scanpy's data loader?

**Comment by esrice** [comment]
Sure, here is an h5 output by cellbender that is parsed fine by scanpy v8 but not v9. Github doesn't allow h5 uploads in comments so here's a download link:

https://oc1.rnet.missouri.edu/index.php/s/fqOxxkjaxkcizUE/download

**Comment by sjfleming** [comment]
Okay thank you @esrice , now I see.  Yes there are still some extra values packed into the h5 file in the cellbender output for v0.2.  I will soon update to v0.3, where the `ptrepack` solution might work.  But for now, that suggestion did not make sense.

The best thing to do is to use the above data loader code for the time being: the `anndata_from_h5()` function.

In v0.3.0, the data loading functions will be distributed as part of cellbender, and also the format of the h5 output files will be tweaked to be compatible with the `scanpy` v1.9+ data loader.

**Comment by HugoCornu** [comment]
thanks for that anndata_from_h5() script !

**Comment by gokceneraslan** [comment]
@sjfleming you might like https://github.com/scverse/scanpy/pull/2344 :)

**Comment by sjfleming** [comment]
Woohoo, thanks @gokceneraslan !

**Comment by carmensandoval** [comment]
Thanks @gokceneraslan ! Is this change available in a public version of Scanpy? I'm on 1.9.1 and still seeing this error.

Name: scanpy
Version: 1.9.1

```
org_1 = sc.read_10x_h5('../cellbender/SAM24425933_cellbender_out_filtered.h5')

Output exceeds the [size limit](command:workbench.action.openSettings?[). Open the full output data [in a text editor](command:workbench.action.openLargeOutput?a8abfd26-9931-48ee-951f-ff7127d3d029)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [6], line 1
----> 1 org_1 = sc.read_10x_h5('../cellbender/SAM24425933_cellbender_out_filtered.h5')

File ~/mambaforge/envs/pegasus/lib/python3.9/site-packages/scanpy/readwrite.py:183, in read_10x_h5(filename, genome, gex_only, backup_url)
    181     v3 = '/matrix' in f
    182 if v3:
--> 183     adata = _read_v3_10x_h5(filename, start=start)
    184     if genome:
    185         if genome not in adata.var['genome'].values:

File ~/mambaforge/envs/pegasus/lib/python3.9/site-packages/scanpy/readwrite.py:268, in _read_v3_10x_h5(filename, start)
    266 try:
    267     dsets = {}
--> 268     _collect_datasets(dsets, f["matrix"])
    270     from scipy.sparse import csr_matrix
    272     M, N = dsets['shape']

File ~/mambaforge/envs/pegasus/lib/python3.9/site-packages/scanpy/readwrite.py:256, in _collect_datasets(dsets, group)
    254 for k, v in group.items():
    255     if isinstance(v, h5py.Dataset):
--> 256         dsets[k] = v[:]
    257     else:
    258         _collect_datasets(dsets, v)
...
---> 86     raise ValueError("Illegal slicing argument for scalar dataspace")
     88 self.mspace = h5s.create(h5s.SCALAR)
     89 self.fspace = fspace

ValueError: Illegal slicing argument for scalar dataspace
```

**Comment by sjfleming** [comment]
Closed by #238
